operator/pkg/ciliumendpointslice/reconciler.go

```diff
@@ -196,13 +218,181 @@ func (r *reconciler) reconcileCESDelete(ces *cilium_v2a1.CiliumEndpointSlice) (e
 }
 
 func (r *reconciler) getCoreEndpointFromStore(cepName CEPName) *cilium_v2a1.CoreCiliumEndpoint {
-	cepObj, exists, err := r.cepStore.GetByKey(cepName.key())
+	podObj, exists, err := r.podStore.GetByKey(cepName.key())
 	if err == nil && exists {
-		return k8s.ConvertCEPToCoreCEP(cepObj)
+		return r.convertPodToCoreCEP(podObj)
 	}
-	r.logger.Debug(fmt.Sprintf("Couldn't get CEP from Store (err=%v, exists=%v)",
+	r.logger.Debug(fmt.Sprintf("Couldn't get Pod from Store (err=%v, exists=%v)",
 		err, exists),
 		logfields.CEPName, cepName.string(),
 	)
 	return nil
 }
+
+// Converts a Pod to a CoreCiliumEndpoint object. Returns nil if no CID has been assigned
+// to the pod.
+func (r *reconciler) convertPodToCoreCEP(pod *slim_corev1.Pod) *cilium_v2a1.CoreCiliumEndpoint {
+	scopedLog := r.logger.With(
+		logfields.K8sPodName, pod.GetName(),
+		logfields.K8sNamespace, pod.Namespace,
+	)
+
+	identityId, err := r.getPodIdentityID(pod)
+	if err != nil {
+		scopedLog.Info("Could not get pod identity ID",
+			logfields.Error, err,
+		)
+		return nil
+	}
+
+	networking, err := GetPodEndpointNetworking(pod)
+	if err != nil {
+		scopedLog.Info("Could not get pod's endpoint networking",
+			logfields.Error, err,
+		)
+		return nil
+	}
+
+	encryptionKey, err := r.getEndpointEncryptionKey(pod)
+	if err != nil {
+		scopedLog.Info("Could not get pod's endpoint encryption key",
+			logfields.Error, err,
+		)
+		return nil
+	}
+
+	namedPorts := r.getNamedPorts(pod)
+
+	return &cilium_v2a1.CoreCiliumEndpoint{
+		Name:       pod.GetName(),
+		IdentityID: identityId,
+		Networking: networking,
+		Encryption: cilium_v2.EncryptionSpec{
+			Key: encryptionKey,
+		},
+		NamedPorts: namedPorts,
+	}
+}
+
+// Get the identity ID for a given pod.
+func (r *reconciler) getPodIdentityID(pod *slim_corev1.Pod) (int64, error) {
+	cid, err := r.getPodIdentity(pod)
+	if err != nil {
+		return 0, err
+	}
+	identityId, err := strconv.ParseInt(cid.Name, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse CID name: %w", err)
+	}
+	return identityId, nil
+}
+
+func (r *reconciler) getPodIdentity(pod *slim_corev1.Pod) (*cilium_v2.CiliumIdentity, error) {
+	k8sLabels, err := ciliumidentity.GetRelevantLabelsForPod(r.logger, pod, r.namespaceStore)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get relevant labels for pod: %w", err)
+	}
+	cidKey := key.GetCIDKeyFromLabels(k8sLabels, labels.LabelSourceK8s)
+	storeCIDs, err := r.cidStore.ByIndex(k8s.ByKeyIndex, cidKey.GetKey())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get CID from store: %w", err)
+	}
+	if len(storeCIDs) == 0 {
+		return nil, fmt.Errorf("CID store is empty")
+	}
+	return storeCIDs[0], nil
```

Comment:
How stable is the order in which entries are returned from the store guaranteed to be? Otherwise, there's the risk of significant churn if a set of labels is associated with two or more identities, depending on which one gets picked at any round.